### PR TITLE
HDDS-8382. Upgrade Ratis to 2.5.1

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Upgrade Ratis to 2.5.1.

Fix for compilation error in tests due to interface change was provided by @szetszwo.  When merging this, please keep the following detail in the commit message:

```
Co-authored-by: Tsz-Wo Nicholas Sze <szetszwo@apache.org>
```

https://issues.apache.org/jira/browse/HDDS-8382

## How was this patch tested?

CI:
https://github.com/adoroszlai/hadoop-ozone/actions/runs/4891132333

### CODE REVIEW

**Comment by adoroszlai** [comment]
`TestDeleteWithSlowFollower` is flaky (HDDS-8430), but it seems to be failing more frequently with new Ratis.

**Comment by adoroszlai** [comment]
With Ratis 2.5.1, block is deleted contrary to test expectations.  This results in two kinds of failures, depending on timing:

https://github.com/apache/ozone/blob/7f7ec2397ac029cf2a8bc8490b7ecafb34020b0d/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestDeleteWithSlowFollower.java#L305-L306

https://github.com/apache/ozone/blob/7f7ec2397ac029cf2a8bc8490b7ecafb34020b0d/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestDeleteWithSlowFollower.java#L307-L317

When block deletion reaches the following condition:

https://github.com/apache/ozone/blob/7f7ec2397ac029cf2a8bc8490b7ecafb34020b0d/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java#L223-L232

with Ratis 2.5.1 `minReplicatedIndex` is always -1 (I guess due to [RATIS-1767](https://issues.apache.org/jira/browse/RATIS-1767), but could be wrong), so deletion is allowed.

I think the `minReplicatedIndex >= 0` condition is wrong.  If `minReplicatedIndex == -1`, then we should not allow deleting the blocks, so I removed that condition.  Test passes locally.  Pushed new commit, waiting for CI in fork.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java [hunk 1]
```diff
@@ -220,7 +220,7 @@ private boolean isDeletionAllowed(ContainerData containerData,
           long minReplicatedIndex =
               ratisServer.getMinReplicatedIndex(pipelineID);
           long containerBCSID = containerData.getBlockCommitSequenceId();
-          if (minReplicatedIndex >= 0 && minReplicatedIndex < containerBCSID) {
```
Comment (smengcl) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java:223]: Is this intended? `minReplicatedIndex` could be `-1`, thus getting rid of the `minReplicatedIndex >= 0` condition here changes the logic slightly.

cc @szetszwo

**Review by smengcl — APPROVED** [review]
lgtm. but I'd prefer another eye on [this](https://github.com/apache/ozone/pull/4664#issuecomment-1537204821).

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java (hunk 1 above)
Comment (szetszwo) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java:223]: I agree that it is correct to remove `minReplicatedIndex >= 0`.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by adoroszlai** [comment]
Thanks @smengcl for the review, @szetszwo for helping out with the state machine tests, reviewing and merging this.

**Pull request merged**